### PR TITLE
test(app-shell): rewrite top level tests to integration and e2e tests

### PR DESCRIPTION
**Pull request by emmenko**
This PR rewrites the top level AppShell tests.

At the moment, the AppShell has enzyme (shallow) tests which are very much tighten to the implementation. As part of our testing strategy migration to RTL, we want to move to refactor the AppShell tests as well.
Doing that comes with some benefits, like less painful and more confident refactorings. This will also help with the migration to TS (#1069).

Testing the AppShell is actually not that trivial. We can't use the test-utils, because they effectively mock out the AppShell, which is what we want to test.
What we can do though is to mock the network requests (_basically the same approach used with the experimental render_).

Long story short: I ended up creating graphql mocks for both RTL and Cypress. The AppShell tests have been rewritten to test the full component and mock the underlying XHR requests.

image

/cc @pa3 for awareness

### CODE REVIEW

**Comment by tdeekens** [comment]
Oh wow. This is really nice 💯. Need to set aside some time to have a closer look. Wondering if it would maybe make sense to announce this in a fe-chapter meeting and pick 1-2 dedicate reviewers how will get a walk high level walkthrough to then review the change?

**Comment by emmenko** [comment]
Sure, I was hoping to get this merged soon-ish so that I can move on with other things.
I can present this in the next fe tasting maybe, although it's really nothing new.

I was also wondering if we would publish a package for those utils, so that we can use them in the apps. We might also include the pulled schemas in the package (still not 100% sure about it though).

**Comment by tdeekens** [comment]
> Sure, I was hoping to get this merged soon-ish so that I can move on with other things.

I see but maybe that's not always possible. Still I think it's not scalable that I am assigned the only reviewer again. Albeit I might have some context. It's not a way we can or should with or towards the team I feel.

**Comment by emmenko** [comment]
Yeah sure, I can ask in slack.

File: cypress/integration/playground/application-shell.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    cy.findByText('Logout').click();
+    const queryParams = encode({
+      reason: LOGOUT_REASONS.USER,
+    });
+    cy.url().should('include', `/logout?${queryParams}`);
+    cy.findByText('This is the logout page for local development.').should(
+      'exist'
+    );
+  });
+  describe('when going to an unknown route', () => {
```
Comment (tdeekens) [inline cypress/integration/playground/application-shell.js]:
```suggestion
  describe('when navigating to an unknown route', () => {
```

File: cypress/integration/playground/application-shell.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    it('should render page not found', () => {
```
Comment (tdeekens) [inline cypress/integration/playground/application-shell.js]:
```suggestion
    it('should render a not found page', () => {
```

File: graphql-test-utils/create-graphql-mock-server.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import { graphql, buildClientSchema } from 'graphql';
+import { addMockFunctionsToSchema } from 'graphql-tools';
+import { formatApolloErrors } from 'apollo-server-errors';
+import loadSchema from './load-schema';
+
+const getRootValue = (operations, operationName, variables) => {
+  if (!operationName || !operations[operationName]) {
+    return {};
+  }
+  const op = operations[operationName];
```
Comment (tdeekens) [inline graphql-test-utils/create-graphql-mock-server.js]: The rootValue is an operation, right?

File: graphql-test-utils/create-graphql-mock-server.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  if (typeof op === 'function') {
+    try {
+      return op(variables);
+    } catch (e) {
+      return e; // properly handle dynamic throw new GraphQLError("message")
+    }
+  }
+  return op;
+};
```
Comment (tdeekens) [inline graphql-test-utils/create-graphql-mock-server.js]: Nit and subjective :) this would have helped me:

```js
const executeGraphQlOperation = (operation, variables) => {
   if (typeof operation === "function") {
      try {
         return operation(variables);
      } catch (e) {
         // TODO: Properly handle dynamic throw new GraphQLError("message")
         return e; 
      }
   }

   return operation
};
const getRootValue = (operations, operationName, variables) => {
   if (!operationName || !operations[operationName]) {
      return {};
   }

   const operation = operations[operationName];

   return executeGraphQlOperation(operation, variables);
};

```

File: graphql-test-utils/create-graphql-mock-server.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+const createGraphqlMockServer = (
+  xhrMock,
+  { mocksByTarget = {}, operationsByTarget = {}, onRequest } = {}
+) => {
+  const processGraphQLRequest = async (targetName, req, res) => {
```
Comment (tdeekens) [inline graphql-test-utils/create-graphql-mock-server.js]: Nit: but we always mix casing.

```suggestion
  const processGraphQlRequest = async (targetName, req, res) => {
```

File: graphql-test-utils/create-graphql-mock-server.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    return res.status(200).body(JSON.stringify(response));
+  };
+  const handleProxyGraphQLRequest = async (req, res) => {
+    const { variables } = JSON.parse(req.body());
+    const targetName = variables.target;
+    return await processGraphQLRequest(targetName, req, res);
+  };
+
+  xhrMock.post('http://localhost:8080/graphql', async (req, res) => {
+    if (onRequest) {
```
Comment (tdeekens) [inline graphql-test-utils/create-graphql-mock-server.js]: When does this happen? Why do we need to wrap the next?

File: graphql-test-utils/create-graphql-mock-server.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      const next = async () => {
+        return await handleProxyGraphQLRequest(req, res);
+      };
+      return await onRequest(req, res, next);
+    }
+    return await handleProxyGraphQLRequest(req, res);
+  });
+  xhrMock.post(/\/api\/graphql$/, async (req, res) => {
```
Comment (tdeekens) [inline graphql-test-utils/create-graphql-mock-server.js]: This is the mc-proxy right? Should we add a comment what requests we're handling maybe?

File: graphql-test-utils/cypress/add-commands.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+Cypress.Commands.add('useMockedGraphqlApi', (customOperations = {}) => {
+  cy.task('getGraphQLSchema', 'mc').then(schema => {
+    cy.mockGraphql({
+      schema,
+    });
+  });
+  const operations = mocksForMc.createMockOperations(customOperations);
+  cy.mockGraphqlOps({ operations });
+  return cy.wrap(operations);
```
Comment (tdeekens) [inline graphql-test-utils/cypress/add-commands.js]: Do you know what this does?

File: graphql-test-utils/graphql-models/user.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  )
+  .attr('launchdarklyTrackingTeam', () => [faker.random.word()])
+  .attr('launchdarklyTrackingTenant', 'gcp-eu')
+  .attr('defaultProjectKey', () => null)
+  .attr('projects', () => ({
+    total: 0,
+    results: [],
+  }));
+
+export default User;
```
Comment (tdeekens) [inline graphql-test-utils/graphql-models/user.js]: Nice that we use rosie and faker here too :)

File: graphql-test-utils/load-schema.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
+const fs = require('fs');
```
Comment (tdeekens) [inline graphql-test-utils/load-schema.js]: Is this really loading? Is it more `resolveAndCacheSchema`?

File: packages/application-shell/src/components/application-shell-provider/utils.spec.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  describe('when the locale is supported by the MC', () => {
+    beforeEach(() => {
+      testWindow = {
+        navigator: {
+          language: 'de-DE',
+        },
+      };
+    });
+    it('should return the locale', () => {
+      expect(getBrowserLocale(testWindow)).toBe('de-DE');
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell-provider/utils.spec.js]: Would it be easier to avoid beforeEach and just do

```js
expect(getBrowserLocale({
  navigator: {
     language: 'de-DE',
  }
})).toBe('de-DE');
```

maybe?

File: packages/application-shell/src/components/navbar/navbar.js [hunk 12]
```diff
@@ -245,7 +245,8 @@ const hasMenuEveryMenuVisibilitySetToBeHidden = (
   Array.isArray(namesOfMenuVisibilities) &&
   namesOfMenuVisibilities.length > 0 &&
   namesOfMenuVisibilities.every(
-    nameOfMenuVisibility => menuVisibilities[nameOfMenuVisibility] === true
+    nameOfMenuVisibility =>
```
Comment (tdeekens) [inline packages/application-shell/src/components/navbar/navbar.js]: What's the reasoning behind the change? I think that's also why the prop becomes non-required, right?

File: packages/application-shell/src/components/application-shell/application-shell.spec.js [hunk 13]
```diff
@@ -1,508 +1,446 @@
...
+    de: { 'CustomApp.title': 'Titel' },
+  },
   ...props,
 });
 
-const testLocaleData = {
-  isLoading: false,
-  locale: 'en',
-  messages: { 'AppKit.title': 'Title en', 'CustomApp.title': 'Title en' },
+const mockSuperfluousRequests = () => {
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell/application-shell.spec.js]: Nit: but should we extract this somehow? Feels like we might have more in the future, right?

File: packages/application-shell/src/components/application-shell/application-shell.spec.js [hunk 14]
```diff
@@ -1,508 +1,446 @@
...
-          })
-        );
-      });
-      it('should not render <ApplicationContextProvider>', () => {
-        expect(wrapper).not.toRender(ApplicationContextProvider);
-      });
+  });
+  it('should not render NavBar', async () => {
+    const rendered = renderApp();
+    await waitForElement(() => rendered.getByText('OK'));
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell/application-shell.spec.js]: Nit I think we can use `await findByText` here everywhere to make it a bit leaner maybe?

File: packages/application-shell/src/components/application-shell/application-shell.spec.js [hunk 15]
```diff
@@ -1,508 +1,446 @@
...
+    createGraphqlMockServer(xhrMock, {
+      operationsByTarget: {
+        mc: mocksForMc.createMockOperations({
+          FetchLoggedInUser: {
+            me: UserMock.build({
+              defaultProjectKey: null,
+              projects: { total: 0, results: [] },
+            }),
+          },
+        }),
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell/application-shell.spec.js]: This is nice!

File: packages/application-shell/src/components/application-shell/application-shell.spec.js [hunk 16]
```diff
@@ -1,508 +1,446 @@
...
-    it('should return the default locale, `en`', () => {
-      expect(getBrowserLocale(testWindow)).toBe('en');
+    await wait(() => {
+      expect(window.location.replace).toHaveBeenCalledWith(
+        `${window.location.origin}/login?${queryParams}`
+      );
+      expect(rendered.queryByText('OK')).not.toBeInTheDocument();
     });
   });
 });
```
Comment (tdeekens) [inline packages/application-shell/src/components/application-shell/application-shell.spec.js]: The test cases are so nice and close to the requirements!

File: graphql-test-utils/create-graphql-mock-server.js (hunk 4 above)
Comment (emmenko) [reply graphql-test-utils/create-graphql-mock-server.js]: Yeah I took this from here: https://github.com/tgriesser/cypress-graphql-mock/blob/6000875ebc9a16f0a3a20fbef72a2c41f0379eca/src/index.ts#L194-L212

I can apply your suggestion though, thanks

File: graphql-test-utils/create-graphql-mock-server.js [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    } catch (error) {
+      // https://github.com/apollographql/apollo-server/blob/61e69412c0aa88b358a2eef4ba635e8618db9946/packages/apollo-server-core/src/runHttpQuery.ts#L295-L297
+      response = { error: formatApolloErrors([error]) };
+    }
+    return res.status(200).body(JSON.stringify(response));
+  };
+  const handleProxyGraphQLRequest = async (req, res) => {
+    const { variables } = JSON.parse(req.body());
+    const targetName = variables.target;
+    return await processGraphQLRequest(targetName, req, res);
```
Comment (emmenko) [inline graphql-test-utils/create-graphql-mock-server.js]:
```suggestion
    return await processGraphQlRequest(targetName, req, res);
```

File: graphql-test-utils/create-graphql-mock-server.js (hunk 6 above)
Comment (emmenko) [reply graphql-test-utils/create-graphql-mock-server.js]: We use it here: https://github.com/commercetools/merchant-center-application-kit/pull/1231/files#diff-935055bd8b0f453f6f0596df2db12c6aR422-R430

This is useful when we want to control the network request, for example failing with 401, which is different from a graphql error.

File: graphql-test-utils/create-graphql-mock-server.js (hunk 7 above)
Comment (emmenko) [reply graphql-test-utils/create-graphql-mock-server.js]: Yes the proxy.

File: graphql-test-utils/cypress/add-commands.js (hunk 8 above)
Comment (emmenko) [reply graphql-test-utils/cypress/add-commands.js]: Yes, it allows to get the value like this:

```js
cy.useMockedGraphqlApi().then(operations => {
})
```

Comment (emmenko) [reply graphql-test-utils/cypress/add-commands.js]: Otherwise it won't work, because Cypress queues everything that happens in the commands and executes them asynchronously.

File: graphql-test-utils/load-schema.js (hunk 10 above)
Comment (emmenko) [reply graphql-test-utils/load-schema.js]: Couldn't find a better name 😅

File: packages/application-shell/src/components/application-shell/application-shell.spec.js (hunk 13 above)
Comment (emmenko) [reply packages/application-shell/src/components/application-shell/application-shell.spec.js]: I guess..should I put it in the graphql-test-utils?

File: packages/application-shell/src/components/application-shell/application-shell.spec.js (hunk 14 above)
Comment (emmenko) [reply packages/application-shell/src/components/application-shell/application-shell.spec.js]: Hmm what do you mean?

File: graphql-test-utils/create-graphql-mock-server.js (hunk 4 above)
Comment (tdeekens) [reply graphql-test-utils/create-graphql-mock-server.js]: Ah kay. Doesn't matter too much. Just what my mind translated the code to :)

File: packages/application-shell/src/components/navbar/navbar.js (hunk 12 above)
Comment (emmenko) [reply packages/application-shell/src/components/navbar/navbar.js]: Yes, it's an optional prop, the prop types were incorrect

File: packages/application-shell/src/components/application-shell/application-shell.spec.js (hunk 13 above)
Comment (tdeekens) [reply packages/application-shell/src/components/application-shell/application-shell.spec.js]: Yeah could help to just have a `applyExternalNetworkRequests` there or something.

File: packages/application-shell/src/components/application-shell/application-shell.spec.js (hunk 14 above)
Comment (tdeekens) [reply packages/application-shell/src/components/application-shell/application-shell.spec.js]: To my understandting RTL released and documented the `findBy`-APIs which are like what we have here so `await waitForElement(() => rendered.getByText('OK'));` can be shortenede to `await rendered.findByText('OK')`.

Comment (emmenko) [reply packages/application-shell/src/components/application-shell/application-shell.spec.js]: Ah, I thought it was only used in cypress or something.

Sure, I we can do that, thanks for the heads up

File: graphql-test-utils/create-graphql-mock-server.js (hunk 6 above)
Comment (tdeekens) [reply graphql-test-utils/create-graphql-mock-server.js]: Ah so it's like intercepting it? I thought it was on any request.

File: graphql-test-utils/cypress/add-commands.js (hunk 8 above)
Comment (tdeekens) [reply graphql-test-utils/cypress/add-commands.js]: Ah this auto queueing promisification. Got it.

File: graphql-test-utils/create-graphql-mock-server.js (hunk 6 above)
Comment (emmenko) [reply graphql-test-utils/create-graphql-mock-server.js]: No only if you pass the `onRequest` option

Comment (tdeekens) [reply graphql-test-utils/create-graphql-mock-server.js]: Yes, it you pass an request interceptor which is called onRequest then it will be intercepted.

File: packages/application-shell/src/test-utils/fixtures/mc.js [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  .attr('owner', {
+    id: faker.random.uuid(),
+  });
+const User = new Factory()
+  .sequence('sequenceId')
+  .attr('email', faker.internet.email())
+  .attr('gravatarHash', faker.random.alphaNumeric(16))
+  .attr('firstName', faker.name.firstName())
+  .attr('lastName', faker.name.lastName())
+  .attr('language', faker.locale)
```
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/application-shell/src/test-utils/fixtures/mc.js]: I want to keep the language to a "known" locale. I can actually just hardcode it to 'en' to avoid confusion.

File: cypress/integration/playground/application-shell.js [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+  //   cy.task('getGraphQLSchema', 'mc').then(schema => {
+  //     cy.mockGraphql({
+  //       schema,
+  //     });
+  //   });
+  // });
+  describe('when user is not authenticated', () => {
+    beforeEach(() => {
+      cy.server();
+      cy.on('window:before:load', win => {
```
[the remark this answers is not in the document]
Comment (emmenko) [reply cypress/integration/playground/application-shell.js]: I tried that first but I wasn't able to make it work.

File: cypress/integration/playground/application-shell.js [hunk 20]
```diff
@@ -70,45 +70,28 @@ describe('when user is authenticated', () => {
...
-        .type(nextProject.name);
-      cy.findByText(nextProject.name).click();
-      cy.url().should('include', `/${nextProject.key}/state-machines`);
+      cy.useMockedGraphqlApi().then(resolvers => {
+        const nextProject = resolvers.FetchLoggedInUser.me.projects.results[1];
+        cy.login();
+        cy.get('[data-test=top-navigation').should('exist');
+        cy.get('input[aria-labelledby="project-switcher"]')
+          .focus()
+          .click({ force: true });
```
[the remark this answers is not in the document]
Comment (emmenko) [reply cypress/integration/playground/application-shell.js]: Somehow it doesn't do anything without `force` 🤷‍♂️

File: packages/application-shell/src/components/application-shell/application-shell.spec.js [hunk 21]
```diff
@@ -1,508 +1,354 @@
...
+  window.localStorage.setItem.mockClear();
+  window.localStorage.getItem.mockClear();
+  window.localStorage.removeItem.mockClear();
+  window.location.replace = jest.fn();
+  reportErrorToSentry.mockClear();
+  window.localStorage.getItem.mockReturnValue('true');
+  ApplicationShellProvider.history = defaultHistory;
+  appShellUtils.getPreviousProjectKey.mockClear();
+});
+afterEach(() => {
```
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/application-shell/src/components/application-shell/application-shell.spec.js]: I would prefer to keep `afterEach`, it gives a better understanding on what needs to be cleaned up.
